### ISSUE TITLE
'Neuron:torch.neuron.trace error: data type not understood' error during compilation

### ISSUE THREAD

**Issue by NatalliaSpichekovaTR** [aws/aws-neuron-sdk#172]
Hi!
While going through a tutorial at https://github.com/aws/aws-neuron-sdk/blob/master/src/examples/pytorch/bert_tutorial/tutorial_pretrained_bert.ipynb, I'm getting 

_WARNING:Neuron:torch.neuron.trace failed on <torch_neuron.graph.Graph object at 0x7fbfc9e8c9b0>; falling back to native python function call
WARNING:Neuron:torch.neuron.trace error: data type not understood_

I've also tryed runnig the resulting code on an inf1.xlarge. The code is running but there are no models loaded on neuron cores and neuron cores are not utilized (I used _neuron-cli list-model_ and _neuron-top_ to check it).

I'm using a custom AMI. Here is info about packages versions:
neuron-cc            1.0.16861.0+6021080341
tensorboard          1.14.0
tensorflow           1.14.0
tensorflow-estimator 1.14.0
torch                1.5.1
torch-neuron         1.5.1.1.0.1532.0

I've also tryed 1.0.18001.0+0.5312e6a21 version of neuron-cc. The behaviour was exactly the same. 
Info on other packages installed as well as my code is in the attached file.

Could you please point me out a possible reason of such behaviour? 

[dummy_bert.pdf](https://github.com/aws/aws-neuron-sdk/files/5276317/dummy_bert.pdf)

**Comment by awsrjh** [comment]
Hi Natallia - let us have a look and get back to you asap.

**Comment by aws-diamant** [comment]
Hey @NatalliaSpichekovaTR,

Thanks for posting the detailed PDF, made it much easier to debug! :)

I did a bit of digging on the error messages that you reported, and I suspect this is caused by a tensorflow/numpy compatibility issues (see: https://github.com/tensorflow/tensorflow/issues/31249). Are you using a virtual environment?

Also, FYI: I just launched a fresh instance and followed the "[Install Neuron in your own AMI](https://github.com/aws/aws-neuron-sdk/blob/master/docs/neuron-install-guide.md#user-guide-configuring-linux-for-repository-updates)" guidelines, and the tutorial seems to work fine for me. So maybe the easiest path forward is to just launch a fresh instance and try again.


Let me know if this works. I'd be happy to help with anything further.

/Ron

**Comment by aws-diamant** [comment]
@NatalliaSpichekovaTR , just checking in -
were you able to overcome this issue?

**Comment by NatalliaSpichekovaTR** [comment]
Thank you for you concern!
I'm in progress of installing necessary enviroment on a fresh ec2. I'll contact you as soon as I finish installation and run compilation on the fresh machine.

**Comment by NatalliaSpichekovaTR** [comment]
@aws-diamant 
Hi!
I've launched a fresh ec2 and installed torch from scratch in accordance with the installation guide you've mentioned.
I've also installed transformers v 3.2.0. But still I'm encountering the same warning 

_WARNING:Neuron:torch.neuron.trace failed on <torch_neuron.graph.Graph object at 0x7f70fd873ac8>; falling back to native python function call
WARNING:Neuron:torch.neuron.trace error: data type not understood_

May I ask you to provide a list of versions of the installed libraries? I'd like to compare your versions vs versions installed on my ec2.

A notebook with code that I used please find in the attachment.
[dummy_bert_v1.pdf](https://github.com/aws/aws-neuron-sdk/files/5290899/dummy_bert_v1.pdf)

**Comment by aws-diamant** [comment]
sure thing, see below.

```
(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show neuron-cc
Name: neuron-cc
Version: 1.0.20600.0+0.b426b885f

(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show numpy
Name: numpy
Version: 1.18.2

(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show tensorflow
Name: tensorflow
Version: 1.15.0

(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show torch
Name: torch
Version: 1.5.1

(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show tensorflow-neuron
Name: tensorflow-neuron
Version: 1.15.3.1.0.2043.0

(test_venv) [ec2-user@ip-172-31-83-12 bert]$ pip show torch-neuron
Name: torch-neuron
Version: 1.5.1.1.0.1721.0
```

I suspect this issue might be related to the versions used by the notebook (I ran this tutorial from the terminal, through a virtual env).
Would you mind trying running the code from the terminal, and see if that makes a difference?
In the meantime, I'll also try to recreate your issue using a notebook. I'll update.

/Ron

**Comment by aws-diamant** [comment]
@NatalliaSpichekovaTR ,

I'm not able to replicate your failure so far, this seem to be working on my end.
Here is the full list of commands I used, starting from a fresh instance (inf1.2xl, Amazon Linux 2 AMI HVM).

```
screen

sudo tee /etc/yum.repos.d/neuron.repo > /dev/null <<EOF
[neuron]
name=Neuron YUM Repository
baseurl=https://yum.repos.neuron.amazonaws.com
enabled=1
metadata_expire=0
EOF

sudo rpm --import https://yum.repos.neuron.amazonaws.com/GPG-PUB-KEY-AMAZON-AWS-NEURON.PUB

sudo yum install aws-neuron-runtime-base

sudo yum install aws-neuron-runtime

sudo yum install aws-neuron-tools

sudo yum update
sudo yum install -y python3 gcc-c++
python3 -m venv test_venv
source test_venv/bin/activate

pip install -U pip

tee $VIRTUAL_ENV/pip.conf > /dev/null <<EOF
[global]
extra-index-url = https://pip.repos.neuron.amazonaws.com
EOF

pip install neuron-cc[tensorflow]

pip install torch-neuron

mkdir bert
cd bert
python3 -m pip install torch-neuron neuron-cc[tensorflow] transformers --upgrade --extra-index-url=https://pip.repos.neuron.amazonaws.com
python bert_compile.py
```

bert_compile.py is taken as-is from the tutorial. also attached for your convenience (with a .txt ext).
[bert_compile.txt](https://github.com/aws/aws-neuron-sdk/files/5296715/bert_compile.txt)

Would you mind giving it a try? I'd like to narrow down to a common starting-point, and continue the debug process from there.

Thanks,
Ron

**Comment by NatalliaSpichekovaTR** [comment]
@aws-diamant 
Hi!
Thank you very much for your assitance!
I tried running a compilation script from console withount installing jupyter notebook in the env. It didn't help - I got the same warning _data type not understood_. What I found out: you have tensorflow v 1.15.0 but in my env I get tensorflow of v 1.14.0. 

I reinstalled neuron with the installation script that you shared in the last comment. I successfully compiled the bert model from the tutorial. I also was able to install jyputer notebook in the same env and compilation worked correctly on both inf1.2xlarge and c5.xlarge ec2s. The compiled code runs on inf1.2xlarge with non-zero utulization of neuron cores.

So, the problem is solved. 
Thank you very much for your effforts and valuable support!
